### ISSUE TITLE
Mac M1: loading libcrypto in an unsafe way

### ISSUE THREAD

**Issue by theothornhill** [cl-plus-ssl/cl-plus-ssl#114]
When trying to install `cl+ssl` on a new M1 mac using quicklisp, it hangs and is pushing me into LDB, showing this error:

```
This is SBCL 2.0.11, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
* (ql:quickload :cl+ssl)
To load "cl+ssl":
  Load 1 ASDF system:
    cl+ssl
; Loading "cl+ssl"
.WARNING: /usr/local/Cellar/sbcl/2.0.11/libexec/bin/sbcl is loading libcrypto in an unsafe way
fatal error encountered in SBCL pid 55608 pthread 0x101042e00:
SIGABRT received.

Welcome to LDB, a low-level debugger for the Lisp runtime environment.
ldb> 
```

From #lisp on freenode, @phoe pointed this out while helping me:
 - https://github.com/cl-plus-ssl/cl-plus-ssl/blob/ff4634a2addedc033394e93a402523532e9b470a/src/reload.lisp#L44-L49
 - https://github.com/orthecreedence/cl-async/issues/181

It seems like a solution is to add a specific version of libcrypto.

**Comment by avodonosov** [comment]
I do not really understand the meaning of the "WARNING: /usr/local/Cellar/sbcl/2.0.11/libexec/bin/sbcl is loading libcrypto in an unsafe way".

But if the problem is about the version of libressl, use code can load any version instead of the default behaviour.
See the :CL+SSL-FOREIGN-LIBS-ALREADY-LOADED description in the doc (https://common-lisp.net/project/cl-plus-ssl/)

Let us know if that helps.

**Comment by phoe** [comment]
> I do not really understand the meaning of the "WARNING: /usr/local/Cellar/sbcl/2.0.11/libexec/bin/sbcl is loading libcrypto in an unsafe way".

Seems like "Clients should not load the unversioned libcrypto dylib as it does not have a stable ABI."

https://stackoverflow.com/questions/58272830/ has some pointers to use a different SSL library binary.

**Comment by theothornhill** [comment]
By reinstalling Homebrew, sbcl and most importantly openssl making sure that in every step we use rosetta 2 it finally works fine.

I guess some parts of that stack was using arm architecture. The `:CL+SSL-FOREIGN-LIBS-ALREADY-LOADED` tip was what finally gave me a wrong architecture stacktrace, so thanks :)

**Comment by fjl** [comment]
It's fixed in my PR.

**Comment by avodonosov** [comment]
@fjl , does the load request for unversioned lib really crashes the process?

@theothornhill  says after fixing his packages everything works for him. So I thought the "unsafe way" message is simply a warning for now, not really breaking the load.

**Comment by theothornhill** [comment]
Yes, it does work for me. However, it seems like a known problem, so I might just be lucky for some reason :)

**Comment by fjl** [comment]
> says after fixing his packages everything works for him.

It will work if cffi finds the library in /usr/local/lib first, or in any of the other non-system locations.

>  So I thought the "unsafe way" message is simply a warning for now, not really breaking the load and crashing the process.

The crash happens when loading "/usr/lib/libcrypto.dylib" because it is a stub whose only purpose is to crash the process. You can find other reports of this, for example: 

- https://stackoverflow.com/questions/58272830/python-crashing-on-macos-10-15-beta-19a582a-with-usr-lib-libcrypto-dylib
- https://developer.apple.com/forums/thread/119429

**Comment by avodonosov** [comment]
Closing this - versioned paths were added in #115

**Comment by ebzzry** [comment]
I encountered the same error on macOS Monterey M1 and my fix was the following:

````
(ql:quickload :cl+ssl/config)
(cl+ssl/config:define-libssl-path "/opt/homebrew/opt/openssl/lib/libssl.dylib")
(cl+ssl/config:define-libcrypto-path "/opt/homebrew/opt/openssl/lib/libcrypto.dylib")
(ql:quickload :cl+ssl)
```

**Comment by ngn999** [comment]
> I encountered the same error on macOS Monterey M1 and my fix was the following:
> 
> ```
> (ql:quickload :cl+ssl/config)
> (cl+ssl/config:define-libssl-path "/opt/homebrew/opt/openssl/lib/libssl.dylib")
> (cl+ssl/config:define-libcrypto-path "/opt/homebrew/opt/openssl/lib/libcrypto.dylib")
> (ql:quickload :cl+ssl)
> ```

I have the same problem. `cl+ssl/config` will work, but my project needs deployed to linux. 

And I found this patch will work:

```
--- a/src/reload.lisp
+++ b/src/reload.lisp
@@ -44,10 +44,10 @@
     ;; so we can just use just "libssl.so".
     ;; More info at https://github.com/cl-plus-ssl/cl-plus-ssl/pull/2.
     (:openbsd "libcrypto.so")
-    (:darwin (:or "/opt/local/lib/libcrypto.dylib"                ;; MacPorts
+    (:darwin (:or "/opt/homebrew/opt/openssl/lib/libcrypto.dylib" ;; Homebrew Arm64
+                  "/opt/local/lib/libcrypto.dylib"                ;; MacPorts
                   "/sw/lib/libcrypto.dylib"                       ;; Fink
                   "/usr/local/opt/openssl/lib/libcrypto.dylib"    ;; Homebrew
-                  "/opt/homebrew/opt/openssl/lib/libcrypto.dylib" ;; Homebrew Arm64
                   "/usr/local/lib/libcrypto.dylib" ;; personalized install
 
                   ;; System-provided libraries. Must be loaded from files with
@@ -92,10 +92,10 @@
     ;; (missing TLSv1_[1,2]_XXX methods,
     ;; see https://github.com/cl-plus-ssl/cl-plus-ssl/issues/56)
     ;; so first try to load possible custom installations of libssl
-    (:darwin (:or "/opt/local/lib/libssl.dylib"                ;; MacPorts
+    (:darwin (:or "/opt/homebrew/opt/openssl/lib/libssl.dylib" ;; Homebrew Arm64
+                  "/opt/local/lib/libssl.dylib"                ;; MacPorts
                   "/sw/lib/libssl.dylib"                       ;; Fink
                   "/usr/local/opt/openssl/lib/libssl.dylib"    ;; Homebrew
-                  "/opt/homebrew/opt/openssl/lib/libssl.dylib" ;; Homebrew Arm64
                   "/usr/local/lib/libssl.dylib" ;; personalized install
 
                   ;; System-provided libraries, with version in the file name.
```

**Comment by fjl** [comment]
Someone needs to dig into the dyld_shared_cache on Monterey and check which libssl versions it supports. I don't have Monterey yet, so can't check.

**Comment by avodonosov** [comment]
The cache seems not playing a role here.  

Moving the Homebrew from the third place to the first place, in front of MacPorts and Fink, fixes the error for @ngn999 and @ebzzry.
So MacPorts or Fink produces the error, while Homebrew does not. Interesting that all the three have no version number in the file name.

**Comment by gustavoFreireS** [comment]
made it work pointing the openssl to `/opt/homebrew/opt/openssl/lib/libcrypto.dylib`

**Comment by avodonosov** [comment]
@gustavoFreireS , pointing how?

**Comment by shukryzablah** [comment]
> I encountered the same error on macOS Monterey M1 and my fix was the following:
> 
> ```
> (ql:quickload :cl+ssl/config)
> (cl+ssl/config:define-libssl-path "/opt/homebrew/opt/openssl/lib/libssl.dylib")
> (cl+ssl/config:define-libcrypto-path "/opt/homebrew/opt/openssl/lib/libcrypto.dylib")
> (ql:quickload :cl+ssl)
> ```

This is the easiest way to do things. I myself had downloaded openssl@3 (normal brew install openssl) and had to use the paths like these: `/usr/local/Cellar/openssl@3/3.0.0_1/lib/libssl.dylib`. Worked perfectly.

**Comment by ngn999** [comment]
> > I encountered the same error on macOS Monterey M1 and my fix was the following:
> > ```
> > (ql:quickload :cl+ssl/config)
> > (cl+ssl/config:define-libssl-path "/opt/homebrew/opt/openssl/lib/libssl.dylib")
> > (cl+ssl/config:define-libcrypto-path "/opt/homebrew/opt/openssl/lib/libcrypto.dylib")
> > (ql:quickload :cl+ssl)
> > ```
> 
> This is the easiest way to do things. I myself had downloaded openssl@3 (normal brew install openssl) and had to use the paths like these: `/usr/local/Cellar/openssl@3/3.0.0_1/lib/libssl.dylib`. Worked perfectly.

can you test my patch?

**Comment by daewok** [comment]
I think I discovered the root cause of the latest troubles. Opened a new issue for it at #138.

**Comment by avodonosov** [comment]
From what @daewok reports in #138, the problem happens when application tries a path tha does not exist.

@ngn999, @shukryzablah, @gustavoFreireS, @ebzzry,  could you confirm that the path tried by cl+ssl on Mac first - /opt/local/lib/libssl.dylib - is absent on your machine?

**Comment by theothornhill** [comment]
FWIW it is absent on my machine. I don't use monterey yet.

**Comment by avodonosov** [comment]
@theothornhill , does the crash reproduce for you? With the question in my previous coment I hope to find confirmation the absence of the file causes the crash.

**Comment by theothornhill** [comment]
@avodonosov Yes, it still crashes now. I had placed the versioned files from `brew install openssl` in `/usr/local/lib` and that made things work.
Now I get:

```
To load "drakma":
  Load 1 ASDF system:
    drakma
; Loading "drakma"
....WARNING: /opt/homebrew/Cellar/sbcl/2.1.3/libexec/bin/sbcl is loading libcrypto in an unsafe way
fatal error encountered in SBCL pid 73807 pthread 0x102c07d40:
SIGABRT received.
```
(it fails on `cl+ssl` when loading drakma)

**Comment by avodonosov** [comment]
@theothornhill, you have the SIGABORT after making the things work by placing the files into `/usr/local/lib`? What bahavior you had before placing the files there?

**Comment by theothornhill** [comment]
Ah sorry I was unclear. After I fixed things by moving the files to `/usr/local/lib` it loads fine, no sigabrt. That was just me cancelling quickloading cl+ssl because it hangs again now after deleting them.

So, before placing files, loading sly hangs in ldb, yielding the message from my first post

**Comment by ngn999** [comment]
> From what @daewok reports in #138, the problem happens when application tries a path that does not exist.
> 
> @ngn999, @shukryzablah, @gustavoFreireS, @ebzzry, could you confirm that the path tried by cl+ssl on Mac first - /opt/local/lib/libssl.dylib - is absent on your machine?

``` bash
$ ls /opt/local/lib/libssl.dylib                           
ls: /opt/local/lib/libssl.dylib: No such file or directory
```

**Comment by ebzzry** [comment]
> > From what @daewok reports in #138, the problem happens when application tries a path that does not exist.
> > @ngn999, @shukryzablah, @gustavoFreireS, @ebzzry, could you confirm that the path tried by cl+ssl on Mac first - /opt/local/lib/libssl.dylib - is absent on your machine?
> 
> ```shell
> $ ls /opt/local/lib/libssl.dylib                           
> ls: /opt/local/lib/libssl.dylib: No such file or directory
> ```

Yes, I can confirm that that path is missing on my machine. I’m using macOS Monterey.

**Comment by avodonosov** [comment]
@theothornhill, ok, I supposed so.

Still, another thing confuses me. I assumed that the absence of `/opt/local/lib/libssl.dylib`, which is the first path in the list of alternatives, causes the crash when cl+ssl tries to load it. But you fix the issue by placing the files to different location - uner `/usr/local/lib`. Strange.

**Comment by daewok** [comment]
> @theothornhill, ok, I supposed so.
> 
> Still, another thing confuses me. I assumed that the absence of `/opt/local/lib/libssl.dylib`, which is the first path in the list of alternatives, causes the crash when cl+ssl tries to load it. But you fix the issue by placing the files to different location - uner `/usr/local/lib`. Strange.

My guess is that it's using DYLD_FALLBACK_LIBRARY_PATH when the requested library is missing. So that would be why /usr/local/lib worked as well.

> Note: If DYLD_FALLBACK_LIBRARY_PATH is not set, dlopen operates as if DYLD_FALLBACK_LIBRARY_PATH was set to $HOME/lib:/usr/local/lib:/usr/lib.

**Comment by avodonosov** [comment]
@daewok, the docs you quoted in #138 say $DYLD_FALLBACK_LIBRARY_PATH is only used when the path does not contain a slash, so this shoul not have influence during the `/opt/local/lib/libssl.dylib` processing.

**Comment by daewok** [comment]
That's why I prefaced those docs with "I'm cautiously optimistic this is a bug". The observations we've made show behavior clearly in contradiction to the manpage.

An even easier test is to run:

```
sbcl --non-interactive --no-userinit --eval '(load-shared-object "/opt/local/lib/libcrypto.dylib")'
```

(or replace the folders in the path to any where there's no libcrypto.dylib present).

On my coworker's machine, the process is then aborted.

**Comment by theothornhill** [comment]
> @theothornhill , could you add `(format t "%load-foreign-library ~A" path)` at https://github.com/cffi/cffi/blob/eac416d11bb483c33c277d164b4c6daccd9af452/src/cffi-scl.lisp#L307 and reproduce the crash?

I cannot seem to get `cffi` to print that line, since it seems to crash before we get there. 

> An even easier test is to run:
> 
> ```
> sbcl --non-interactive --no-userinit --eval '(load-shared-object "/opt/local/lib/libcrypto.dylib")'
> ```

This reproduces for me too.

**Comment by avodonosov** [comment]
@theothornhill , strange that you can't get the line printed. Maybe remove .fasl files and rebuild (that's an unlined function). On Linux .fals files can be removed by `rm ~/.cache/common-lisp/`

**Comment by daewok** [comment]
@theothornhill @avodonosov I think the wrong line to edit was mentioned. It should be <https://github.com/cffi/cffi/blob/eac416d11bb483c33c277d164b4c6daccd9af452/src/cffi-sbcl.lisp#L366> (cffi-sbcl.lisp instead of cffi-scl.lisp)

**Comment by theothornhill** [comment]
Thanks, I wondered if that could be the case. I'll check tonight :)

**Comment by theothornhill** [comment]
@avodonosov @daewok I'm sorry, but I cannot get it to work... I've tried to quickload `cffi` before quickloading `cl+ssl`, and also with manually recompiling that function, and the functions it is referenced from, since it's an inlined function.

What am I missing?

**Comment by avodonosov** [comment]
Difficwlt to say what is missing...

**Comment by aniketnarvekar** [comment]
> @theothornhill @avodonosov I think the wrong line to edit was mentioned. It should be https://github.com/cffi/cffi/blob/eac416d11bb483c33c277d164b4c6daccd9af452/src/cffi-sbcl.lisp#L366 (cffi-sbcl.lisp instead of cffi-scl.lisp)

Using your reference I have fixed the issue which was:-
SBCL attempted to load libraries that were not present in the path leading to the crash.
I fixed the problem on cffi library changes are [https://github.com/cffi/cffi/pull/193/files](here).

**Comment by aniketnarvekar** [comment]
> @theothornhill @avodonosov I think the wrong line to edit was mentioned. It should be https://github.com/cffi/cffi/blob/eac416d11bb483c33c277d164b4c6daccd9af452/src/cffi-sbcl.lisp#L366 (cffi-sbcl.lisp instead of cffi-scl.lisp)
Using your reference I found out that 'try-foreign-library-alternatives' function in libraries.lisp is loading each library from the library-list parameter without checking if its present at specified location. [Here](https://github.com/cffi/cffi/pull/193) is my solution. For me this fixes the sbcl crashes while loading libcrypto library.

**Comment by aniketnarvekar** [comment]
> @theothornhill @avodonosov I think the wrong line to edit was mentioned. It should be https://github.com/cffi/cffi/blob/eac416d11bb483c33c277d164b4c6daccd9af452/src/cffi-sbcl.lisp#L366 (cffi-sbcl.lisp instead of cffi-scl.lisp)

Using your reference I found out that 'try-foreign-library-alternatives' function in libraries.lisp is loading each library from the library-list parameter without checking if its present at specified location. Here is my solution. For me this fixes the sbcl crashes while loading libcrypto library.

**Comment by daewok** [comment]
@theothornhill Maybe you can `(trace sb-alien:load-shared-object)` instead.

@aniketnarvekar That is not a good solution. It looks like your patch is at <https://github.com/cffi/cffi/pull/193> (it was mentioned in an earlier comment, not sure why it was deleted).

First, it will not work when loading a system library. You typically do not provide the full path to those, so the `probe-file` will return NIL except in the unlikely event that `*default-pathname-defaults*` is bound to the directory containing the library.

Second, Apple has decided to remove ~all system libraries from the file system and store them in a cache elsewhere. So `probe-file` will *never* return non-NIL for those. Apple says the correct way to test for library existence is to `dlopen` them and check for an error. That makes this bug extra rich since it seems that `dlopen("/some/non-existent/path/libcrypto.dylib")` causes the process to abort instead of returning an error.

**Comment by avodonosov** [comment]
Anyone sees a problem if I remove all non-standard paths for Darwin - leaving only ‘/usr/lib/libssl.NN.dylib‘?

**Comment by daewok** [comment]
> Anyone sees a problem if I remove all non-standard paths for Darwin - leaving only /usr/lib/libssl.NN.dylib ?

Seems reasonable, especially since `cl+ssl/config` was introduced.

**Comment by jaemk** [comment]
Others above already mentioned that moving files to `/usr/local/lib` fixed it, but here's some explicit steps for anyone else struggling with this on osx (I hit this after updating to Monterey),
- `brew install openssl` (probably installed to `/usr/local/opt/openssl@3`, or a higher `@`)
- `brew link openssl` _doesn't work_, so you need to symlink this custom openssl install manually so cl+ssl can find it
- based on https://github.com/fjl/cl-plus-ssl/blob/3aaba3feb48e0e23b09c46cc7d51a89868e2ed80/src/reload.lisp#L47 it'll look for `/usr/local/lib/libssl.dylib` / `libcrypto.dylib`
- so we need to link to the above (after installing with brew):

In total:
```
brew install openssl
ln -s /usr/local/opt/openssl\@3/lib/libcrypto.dylib /usr/local/lib/libcrypto.dylib
ln -s /usr/local/opt/openssl\@3/lib/libssl.dylib /usr/local/lib/libssl.dylib
```

**Comment by avodonosov** [comment]
@aniketnarvekar, thank you for confirming a non-existing path crashes the app. I was actually curious to know exactly what path it is - is it one of the path specified in reload.lisp or cffi also falls-back to some default path which causes the crash (reading the cffi code I think it should not, but was curious to be sure).

As for checking the path presence, while @daewok is right that a generic cffi patch is not good for that, we can do that for our specific paths on cl+ssl level. In the commit referenced above I implemented that.

So now we try custom paths only if they are present. I hesitated to remove them altogether - do not want to cause inconvenience for people who may depend on them.

For the record, the custom paths where added in #56.
